java/arcs/core/analysis/DependencyNode.kt

```diff
@@ -0,0 +1,131 @@
+private typealias Identifier = String
+
+/** Lists of [Identifier]s imply an [AccessPath]. */
+private typealias Path = List<Identifier>
+
+/**
+ * [DependencyNode]s represent how inputs of a Paxel [Expression] contribute to its outputs.
+ *
+ * A [DependencyNode] makes up a Directed-Acyclic-Graph to map fields of an output handle
+ * connection to fields in other handle connections.
+ *
+ * - [DependencyNode.PrimitiveValue] represents an input handle connection and access path.
+ * - [DependencyNode.DerivedFrom] indicates that an input has been modified in the Paxel expression.
+ * - [DependencyNode.AggregateValue] connects partial access paths to other nodes in the graph.
+ *   These are used to form left-hand-side / right-hand-side relations between handle connections.
+ *
+ * Example:
+ *   ```
+ *   particle HousePets
+ *     input: reads PetData { cats: Number, dogs: Number, household: Text }
+ *     output: writes PetStats {
+ *       ratio: inline Domestication {trained: Number, total: Number},
+ *       family: Text,
+ *       limit: Number
+ *     } = new PetStats {
+ *       ratio: new Domestication {
+ *         trained: input.cats,
+ *         total: input.cats + input.dogs,
+ *       },
+ *       family: input.household,
+ *       limit: 5
+ *     }
+ *   ```
+ *
+ *   This can be translated to:
+ *
+ *   ```
+ *   DependencyNode.AggregateValue(
+ *    "ratio" to DependencyNode.AggregateValue(
+ *      "trained" to DependencyNode.PrimitiveValue("input", "cats"),
+ *      "total" to DependencyNode.DerivedFrom(
+ *        DependencyNode.PrimitiveValue("input", "cats"),
+ *        DependencyNode.PrimitiveValue("input", "dogs")
+ *      )
+ *    ),
+ *    "family" to DependencyNode.PrimitiveValue("input", "household"),
+ *    "limit" to DependencyNode.LITERAL
+ *   )
+ *   ```
+ *
+ *   We can represent this DAG graphically:
+ *
+ *                    (root)
+ *            __________|______
+ *           /      |          \
+ *          /      /        _(ratio)_
+ *        /       /        |         \
+ *   (limit)  (family)  (trained)  (total)
+ *     |        |          |     /   |
+ *     |        |          |   /     |
+ *     x     (household)  (cats)  (dogs)
+ *
+ *   This internal representation, in turn, can be translated into the following claims:
+ *
+ *   ```
+ *   claim output.ratio.trained derives from input.cats
+ *   claim output.ratio.total derives from input.cats and derives from input.dogs
+ *   claim output.family derives from input.household
+ *   ```
+ */
+sealed class DependencyNode {
+
+  /** An unmodified input (from a handle connection) used in a Paxel [Expression]. */
+  data class PrimitiveValue(val path: Path = emptyList()) : DependencyNode() {
+    constructor(vararg fields: Identifier) : this(listOf(*fields))
+  }
+
+  /** Represents modification of an access path within a Paxel [Expression]. */
+  data class DerivedFrom(val inputs: Set<DependencyNode> = emptySet()) : DependencyNode() {
```

Comment:
I made the constructor private, however doing so causes issues with the data class: There's a warning that this property will still be exposed via the generated `copy` method. Because of this, I opted to make the primary constructor public. Do you think this is a concern? If not, I'll happily make it private. 

> Should this ever be an emptySet?

We use the concept of an empty `DerivedFrom` to represent literals right now. This in vestigial from when we had `union` / `plus` that promoted input nodes to derive nodes.

Reply:
If it's a `Set<Path>`, I don't think the ctor would need to be private. WDYT?

Reply:
Let us make it `Set<PrimitiveValue>` (or `Set<Input>` if you make rename `PrimitiveValue`)